NAS-108501 / 12.0 / samba - fix shadow copy browsing

anodos325 opened the pull request:
TrueNAS 12.0 expanded struct smb_filename to include a timeval
struct vss_tval that is populated during unix_convert() in filename.c.
This allows simpler checks for whether a file is in a snapshot directory
(simply checking whether tv_sec > 0), and significantly reduces impact
of having shadow copies enabled. There were a few places inside
smbd and vfs modules where synthetic_smb_fname() was called and
the vss_tval struct wasn't copied into the new smb_filename struct
resulting in subsequent calls to check_name() / realpath() / stat()
to fail or give surprising results.

Backport statcache changes from 4.13 to account for VSS timestamps
in struct smb_filename. Ensure that VSS timestamp is added after
synthetic_smb_fname() calls where needed.

anodos325 committed:
s3:modules:smblibzfs - synchronize with master

Bring in API improvements from Master. Fix leak of struct smb_zhandle
during repeated SMB tree connects on same path. Lazy initialize
list of dataset information during SMB2 tree-connect, which is reused
on subsequent tree-connects. This lazy-initializes dataset / libzfs
info and should reduce server load during repeated tree connects.

anodos325 committed:
s3:modules:shadow_copy_zfs - prepend SNUM to snapcache entries

Shares now have global cache for shadow path to unix path lookups.
Ensure we prepend SNUM to the keys to avoid collisions between shares.

anodos325 committed:
s3:modules:shadow_copy_zfs - properly handle VSS relative paths

If we've already changed our working directory through vfs_chdir_fn()
to a snapdir, then subsequent path / vss lookups will contain
the snapshot mountpoint once we resolve_path(). This needs to be
accounted for in our conversion for @GMT strings to paths on the
server.

anodos325 committed:
samba - plumb vss_tval through Samba's VFS

TrueNAS 12.0 expanded struct smb_filename to include a timeval
struct vss_tval that is populated during unix_convert() in filename.c.
This allows simpler checks for whether a file is in a snapshot directory
(simply checking whether tv_sec > 0), and significantly reduces impact
of having shadow copies enabled. There were a few places inside
smbd and vfs modules where synthetic_smb_fname() was called and
the vss_tval struct wasn't copied into the new smb_filename struct
resulting in subsequent calls to check_name() / realpath() / stat()
to fail or give surprising results.

Backport statcache changes from 4.13 to account for VSS timestamps
in struct smb_filename. Ensure that VSS timestamp is added after
synthetic_smb_fname() calls where needed.

william-gr commented on net/samba/files/0001-add-ix-custom-vfs-modules.patch:
```diff
@@ -509,64 +523,234 @@ index 00000000000..6ffaab09bdf
-+	char nd[PATH_MAX] = { 0 };
++	char *slashp = NULL;
++	*nslashes = 0;
++	strlcpy(buf, path, buflen);
```
I imagine the path is sanitized at this point? e.g. no /mnt/foo///bar///user?

anodos325 replied:
Path is sanitized prior to here in source3/smbd/filename_convert.c

william-gr commented on net/samba/files/0001-add-ix-custom-vfs-modules.patch:
```diff
@@ -509,64 +523,234 @@ index 00000000000..6ffaab09bdf
++	char *parent_dsname = NULL;
++
++	parent_dsname = zfs_get_name(zfsp);
++	rv = zfs_prop_get(zfsp, ZFS_PROP_MOUNTPOINT, parent_mp,
```
Isnt this supposed to be &parent_mp or am I crazy? probably the latter

william-gr commented on net/samba/files/0001-add-ix-custom-vfs-modules.patch:
```diff
@@ -509,64 +523,234 @@ index 00000000000..6ffaab09bdf
 +		return -1;
 +	}
-+	if (parent == NULL) {
-+		return (-1);
```
What if the mountpoint is not set? e.g. "legacy" Not possible here?

william-gr commented on net/samba/files/0001-add-ix-custom-vfs-modules.patch:
```diff
@@ -509,64 +523,234 @@ index 00000000000..6ffaab09bdf
++		return -1;
++	}
++	rv = zfs_mount(new, NULL, 0);
++	if (rv != 0) {
```
Dont we have to close new first? Same later on quote set?

anodos325 committed:
s3:modules:smblibzfs - improve error handling on create_dataset()

On failure to mount dataset, immediately close zhandle and return
error to prevent subsequent quota set command from covering up
significant error condition.
